src/plugins/arithmetics/add_operation.cpp

```diff
@@ -342,32 +389,302 @@ namespace phylanx { namespace execution_tree { namespace primitives
         auto cv = rhs.vector();
         auto cm = lhs.matrix();
 
-        if (cv.size() != cm.columns())
+        if (cv.size() == cm.rows())
         {
-            HPX_THROW_EXCEPTION(hpx::bad_parameter,
-                "add_operation::add2d1d",
-                execution_tree::generate_error_message(
-                    "vector size does not match number of matrix columns",
-                    name_, codename_));
+            if (lhs.is_ref())
+            {
+                blaze::DynamicMatrix<double> m{cm.rows(), cm.columns()};
+                for (std::size_t i = 0; i != cm.columns(); ++i)
+                {
+                    blaze::column(m, i) = blaze::column(cm, i) + cv;
+                }
+                return primitive_argument_type{std::move(m)};
+            }
+            else
+            {
+                for (std::size_t i = 0; i != cm.columns(); ++i)
+                {
+                    blaze::column(cm, i) += cv;
+                }
+
+                return primitive_argument_type{std::move(lhs)};
+            }
+        }
+        else if (cv.size() == cm.columns())
+        {
+            if (lhs.is_ref())
+            {
+                blaze::DynamicMatrix<double> m{cm.rows(), cm.columns()};
+                for (std::size_t i = 0; i != cm.rows(); ++i)
+                {
+                    blaze::row(m, i) = blaze::row(cm, i) + blaze::trans(cv);
+                }
+                return primitive_argument_type{std::move(m)};
+            }
+            else
+            {
+                for (std::size_t i = 0; i != cm.rows(); ++i)
+                {
+                    blaze::row(cm, i) += blaze::trans(cv);
+                }
+
+                return primitive_argument_type{std::move(lhs)};
+            }
+        }
+        HPX_THROW_EXCEPTION(hpx::bad_parameter,
+            "add_operation::add2d1d",
+            execution_tree::generate_error_message(
+                "vector size does not match either the number of matrix "
+                "columns nor rows.",
+                name_, codename_));
+    }
+
+    int add_operation::get_stretch_dimension(
+        std::size_t lhs, std::size_t rhs) const
+    {
+        // 0: No copy, 1: stretch lhs, 2: stretch rhs
+        if (lhs != rhs)
+        {
+            // lhs x dimension must be stretched
+            if (lhs == 1)
+            {
+                return 1;
+            }
+            // rhs x dimensions must be stretched
+            else if (rhs == 1)
+            {
+                return 2;
+            }
         }
+        return 0;
+    }
```

Comment:
I'd suggest using an enumerator type instead of integer values for distinguishing the necessary operations.